Migrate api section to 6 0 (#11687)
storybookjs/storybook: docs/api/addons.md

---
title: 'Addons'
---

Addons extend Storybook with features and integrations that are not built into the core. Most Storybook features are implemented as addons. For instance: documentation, accessibility testing, interactive controls, and design previews. 
Storybook’s addon API makes it easy for you to configure and customize Storybook in new ways. There are countless addons made by the community that unlock time-saving workflows.  What addons can do:

- Add a panel to Storybook (like Action Logger).
- Add a tool to Storybook’s toolbar (like zoom or grid).

> - Add a tab to Storybook (like SB Docs).

Author: I'm ok with that, we'll need to discuss a bit more about where to point them at.

> - Add a tab to Storybook (like SB Docs).

Reviewer: Oh, right so the first two to the sections inside essentials, and the third to the "writing docs" section.

> - Add a tab to Storybook (like SB Docs).

Author: Added the link, let me know if it's ok.

- Interact and communicate with other addons or Storybook UI
- Change Storybook’s state using its APIs.
- Navigate within Storybook.

> - Register keyboard shortcuts (coming soon).

Author: Updated per feedback. Marking this one as closed

---
title: 'Addons'
---

Addons extend Storybook with features and integrations that are not built into the core. Most Storybook features are implemented as addons. For instance: documentation, accessibility testing, interactive controls, and design previews. 

> Storybook’s addon API makes it easy for you to configure and customize Storybook in new ways. There are countless addons made by the community that unlock time-saving workflows.  What addons can do:

Reviewer 2: ```suggestion
The addon API makes it easy to configure and customize Storybook in new ways. There are countless addons made by the community that unlock time-saving workflows.  What addons can do:
```

Author: Fixed with the latest commit.

- [Add a panel to Storybook (like Action Logger).](../essentials/actions)
- [Add a tool to Storybook’s toolbar (like zoom or grid).](../essentials/toolbars-and-globals)
- [Add a tab to Storybook (like SB Docs).](../writing-docs/introduction)

Browse the [Addon gallery](/addons) to install an existing addon or as inspiration for your own addon. Read on to learn how to make an addon yourself.

### Storybook basics

Before writing your first addon, let’s take a look at the basics of Storybook’s architecture. While Storybook presents a unified user interface, under the hood it’s divided down the middle into **Manager** and **Preview**.

The Manager is the UI where Storybook’s search, navigation, toolbars, and addons are rendered. The Preview area is an iframe where stories are rendered.

![Storybook detailed window](./manager-preview.jpg)

> Because Manager and Preview run in separate iframes, they communicate across a communication channel. For example, when you select a story within the Manager, an event is sent across the channel, and the selected story is rendered inside the Preview.

Reviewer 2: ```suggestion
Because Manager and Preview run in separate iframes, they communicate across a communication channel. When you select a story within the Manager an event is sent across the channel and the selected story is rendered inside the Preview.
```

Author: Fixed with latest commit

import { AddonPanel } from '@storybook/components';

const ADDON_ID = 'myaddon';
const PANEL_ID = `${ADDON_ID}/panel`;

const MyPanel = () => <div>MyAddon</div>;
addons.register(ADDON_ID, api => {
  addons.add(PANEL_ID, {
    type: types.PANEL,
    title: 'My Addon',
    render: ({ active, key }) => (
      <AddonPanel active={active} key={key}>
        <MyPanel />
      </AddonPanel>
    )
  });
});
```

This is boilerplate code for any addon that adds a panel to Storybook, and there’s really not much going on here. In this case, we’re just adding a static div that renders when the panel is selected in Storybook’s UI.

#### Display story parameter

Next, let’s replace the `MyPanel` component from above to show the parameter.

```js
// register.js
import { useParameter } from '@storybook/api';
const PARAM_KEY = 'myAddon';
const MyPanel = () => {
  const value = useParameter(PARAM_KEY, null);
  const item = value ? value.data : "";
  return <div>{item}</div>;
}
```

The new version is made smarter by `useParameter`, which is a [React hook](https://reactjs.org/docs/hooks-intro.html) that updates the parameter value and re-renders the panel every time the story changes.

> Storybook’s addons API provides hooks like this so all of that communication can happen behind the scenes and you can focus on your addon functionality.

Reviewer 2: ```suggestion
The addon API provides hooks like this so all of that communication can happen behind the scenes. That means you can focus on your addon's functionality.
```

Author: Fixed with latest commit

#### Register the addon

Finally, let’s hook it all up. Addons are typically published as standalone packages, but they can also be written locally in an existing Storybook project. We’ll make our addon a local addon.

> In a Storybook project, update your [`.storybook/main.js`](../configure/overview#configure-story-rendering):

Reviewer 2: ```suggestion
Update your [`.storybook/main.js`](../configure/overview#configure-story-rendering):
```

Author: Fixed with latest commit

```js
// .storybook/main.js
module.exports = {
  addons: ['path/to/register.js']
}
```
The path can be an absolute location on your file system, or a path relative to your `.storybook` directory (e.g. `./my-addon/register.js` if you defined the addon inside your `.storybook` folder).

If you get an error similar to:

```sh
ModuleParseError: Module parse failed: Unexpected token (92:22)

You may need an appropriate loader to handle this file type.
       var value = this.state.value;
      var active = this.props.active;
      return active ? <div>{value}</div> : null;
     }
   }]);
```
It is likely because you do not have a `.babelrc` file or do not have it configured with the correct presets:

```json
{ 
    "presets": ["@babel/preset-env", "@babel/preset-react"]
}
```
 

> Now restart/rebuild storybook and your addon should appear in the addons panel. Furthermore, as you navigate between stories, the parameter displayed should update accordingly.

Reviewer 2: ```suggestion
Now restart/rebuild Storybook and your addon should appear in the addons panel. Furthermore, as you navigate between stories, the parameter displayed should update accordingly.
```

Author: Fixed with latest commit

### Addon recipes

Once you understand the basics of writing an addons, there are a variety of common enhancements to make your addon better.

#### Disabling the addon panel

It’s possible to disable the addon panel for a particular story.

To make that possible, you need to pass the `paramKey` element when you register the panel:

```js
addons.register(ADDON_ID, () => {
  addons.add(PANEL_ID, {
    type: types.PANEL,
    title: 'My Addon',
    render: () => <div>Addon tab content</div>,
    paramKey: 'myAddon', // this element
  });
});
```
Then when adding a story, you can then pass a disabled parameter.

```js
// Button.story.js

import React from 'react';
export default {
  title: 'Button',
  parameters: {
    myAddon: { disable: true },
  },
};
```

#### Styling your addon

Storybook uses [Emotion](https://emotion.sh/docs/introduction) for styling, AND we provide a theme which can be set by the user!

> We recommend you also use Emotion to style your addon’s components. If you use Emotion, you can use the active storybook theme, which benefits users.

Reviewer 2: ```suggestion
We recommend you also use Emotion to style your addon’s UI components. That allows you to use the active Storybook theme to deliver a seamless developer experience.
```

Author: Fixed with latest commit

If you don’t want to use Emotion, you can use inline styles or another css-in-js lib. You can receive the theme as a prop by using the `withTheme` hoc from Emotion. [Read more about theming](../configure/user-interface#theming).

#### Storybook components

> You can write your addon UI using any React library, but to make writing addons easier, we’ve published Storybook’s UI components for your reuse as `@storybook/components`. When you use Storybook components you get:

Reviewer 2: ```suggestion
Addon authors can develop their UIs using any React library. But we recommend using Storybook’s own UI components in `@storybook/components` to build addons faster. When you use Storybook components you get:
```

Author: Fixed with latest commit

- Battled tested off-the-shelf components
- Storybook native look and feel
- Built-in support for Storybook theming

You can check them out in [Storybook’s own storybook](https://storybookjs.netlify.app/) 

#### Packaging

> In the example above, we showed how to write a local addon inside an existing Storybook project. To distribute your addon more broadly, you can package the addon into a standalone NPM module.

Reviewer 2: ```suggestion
In the example above, we showed how to write a local addon inside an existing Storybook project. To distribute your addon for others, package the addon into a standalone NPM module.
```

Author: Fixed with latest commit
